File: baixar-e-solicitar-xml-sefaz/src/upload/uploader.py
"""
Função principal para upload automático de XMLs para o SIEG
"""
import os
import time
from typing import Optional, Dict
from concurrent.futures import ThreadPoolExecutor, as_completed

from src.core.config import PATHS, SIEG_API_KEY
from src.upload.utils import validar_xml, extrair_chave_acesso, identificar_tipo_xml, obter_tipo_completo_nota
from src.upload.sieg_api import enviar_xml


# Configurações de warm-up (aquecimento gradual)
WARM_UP_FASE1_QTD = 5       # Quantidade de XMLs na fase 1 (sequencial)
WARM_UP_FASE1_THREADS = 1   # Threads na fase 1
WARM_UP_FASE2_QTD = 15      # Quantidade de XMLs na fase 2
WARM_UP_FASE2_THREADS = 5   # Threads na fase 2
WARM_UP_FASE3_QTD = 30      # Quantidade de XMLs na fase 3
WARM_UP_FASE3_THREADS = 10  # Threads na fase 3
WARM_UP_DELAY = 0.2         # Delay entre envios no warm-up (segundos)

# Configurações padrão
NUM_THREADS_PADRAO = 20


def processar_xml_individual(xml_info: Dict, api_key: Optional[str] = None) -> tuple:
    """
    Processa um XML individual: valida e envia diretamente
    
    Args:
        xml_info: Dict com informações do XML (caminho, nome, conteudo, chave, tipo)
        api_key: API Key do SIEG (usa do config se None)
        
    Returns:
        Tupla (sucesso: bool, ja_existia: bool, mensagem: str)
    """
    if not api_key:
        api_key = SIEG_API_KEY
    
    if not api_key:
        return False, False, "API Key não configurada"
    
    nome = xml_info.get('nome', 'Desconhecido')
    conteudo = xml_info.get('conteudo', '')
    
    # Valida XML
    if not validar_xml(conteudo):
        return False, False, "XML inválido"
    
    # Envia XML
    sucesso, msg, _ = enviar_xml(conteudo, api_key, silencioso=True)
    
    if sucesso:
        return True, False, "Enviado com sucesso"
    else:
        return False, False, msg


def enviar_automatico(
    pasta: Optional[str] = None,
    excluir_enviados: bool = True,
    num_threads: int = NUM_THREADS_PADRAO
) -> Dict:
    """
    Modo automático: valida, envia e exclui XMLs sem interação.
    Usa warm-up gradual para evitar sobrecarga no servidor.
    
    Args:
        pasta: Pasta com XMLs (usa PATHS.downloads_dir se None)
        excluir_enviados: Se True, exclui XMLs enviados com sucesso
        num_threads: Número máximo de threads (após warm-up)
        
    Returns:
        Dict com estatísticas do envio
    """
    print("="*60)
    print("MODO AUTOMÁTICO - Envio de XMLs para SIEG")
    print("="*60)
    
    # Usar pasta padrão se não informada
    if not pasta:
        pasta = str(PATHS.downloads_dir)
    
    print(f"\n📁 Pasta: {pasta}")
    print(f"🔧 Threads máx: {num_threads}")
    print(f"🗑️  Excluir após envio: {'Sim' if excluir_enviados else 'Não'}")
    print()
    
    # Verificar se pasta existe
    if not os.path.exists(pasta):
        print(f"\nERRO: Pasta não encontrada: {pasta}")
        return {"erro": "Pasta não encontrada"}
    
    # Obter API Key
    api_key = SIEG_API_KEY
    if not api_key:
        print("\nERRO: API Key não configurada no .env")
        return {"erro": "API Key não configurada"}
    
    # Buscar XMLs
    print(f"\n{'='*60}")
    print("Buscando arquivos XML...")
    print("="*60)
    
    xmls_encontrados = []
    for root_dir, dirs, files in os.walk(pasta):
        for arquivo in files:
            if arquivo.lower().endswith('.xml'):
                caminho_completo = os.path.join(root_dir, arquivo)
                xmls_encontrados.append(caminho_completo)
    
    if not xmls_encontrados:
        print("ℹ️  Nenhum arquivo XML encontrado na pasta.")
        return {"total": 0, "enviados": 0, "existentes": 0, "erros": 0}
    
    print(f"📦 Encontrados: {len(xmls_encontrados)} arquivo(s) XML")
    
    # Ler e validar XMLs
    print(f"\n{'='*60}")
    print("🔍 Lendo e validando XMLs...")
    print("="*60)
    
    xmls_validos = []
    for caminho in xmls_encontrados:
        try:
            with open(caminho, 'r', encoding='utf-8') as f:
                xml_content = f.read()
            
            if not validar_xml(xml_content):
                continue
            
            chave_acesso = extrair_chave_acesso(xml_content)
            tipo_xml = identificar_tipo_xml(xml_content)
            xmls_validos.append({
                'caminho': caminho,
                'nome': os.path.basename(caminho),
                'conteudo': xml_content,
                'chave': chave_acesso,
                'tipo': tipo_xml
            })
        except:
            continue
    
    print(f"✅ XMLs válidos: {len(xmls_validos)}")
    
    if not xmls_validos:
        print("\nℹ️  Nenhum XML válido encontrado.")
        return {"total": 0, "enviados": 0, "existentes": 0, "erros": 0}
    
    # Enviar todos os XMLs validos diretamente, sem consulta previa de existencia.
    print(f"\n{'='*60}")
    print("Preparando XMLs para envio direto ao SIEG...")
    print("="*60)
    print("Consulta previa de existencia desativada; todos os XMLs validos serao enviados.")
    
    xmls_para_enviar = xmls_validos
    xmls_ja_existentes = []
    print(f"   Para enviar: {len(xmls_para_enviar)}")
    
    # ENVIO COM WARM-UP
    print(f"\n{'='*60}")
    print("🔥 Iniciando envio com WARM-UP gradual...")
    print("="*60)
    print(f"   📤 Fase 1: {WARM_UP_FASE1_QTD} XMLs com {WARM_UP_FASE1_THREADS} thread(s) (sequencial)")
    print(f"   📤 Fase 2: {WARM_UP_FASE2_QTD} XMLs com {WARM_UP_FASE2_THREADS} thread(s)")
    print(f"   📤 Fase 3: {WARM_UP_FASE3_QTD} XMLs com {WARM_UP_FASE3_THREADS} thread(s)")
    print(f"   📤 Fase 4: Restante com {num_threads} thread(s)")
    print(f"\n⏳ Enviando {len(xmls_para_enviar)} XML(s) para o SIEG...\n")
    
    enviados_sucesso = []
    erros_envio = []
    erros_recuperaveis = []  # Para retentar no final
    
    inicio_tempo = time.time()
    total_para_enviar = len(xmls_para_enviar)
    processados = 0
    
    def enviar_com_progresso(xml_info, fase_info=""):
        nonlocal processados
        sucesso, msg, recuperavel = enviar_xml(xml_info['conteudo'], api_key, silencioso=True)
        
        processados += 1
        
        # Obtém chave completa (44 dígitos) e tipo completo com código
        chave = xml_info.get('chave', '')
        tipo_completo = obter_tipo_completo_nota(xml_info['conteudo'])
        
        # Formata chave: mostra completa se tiver 44 dígitos
        if chave and len(chave) == 44:
            chave_formatada = f"N_{chave}"
        else:
            # Se não tem chave completa, tenta extrair novamente do conteúdo
            chave_extraida = extrair_chave_acesso(xml_info['conteudo'])
            if chave_extraida and len(chave_extraida) == 44:
                chave_formatada = f"N_{chave_extraida}"
            else:
                # Fallback: usa nome do arquivo truncado
                chave_formatada = xml_info['nome'][:45] + "..."
        
        tipo_info = f" ({tipo_completo})" if tipo_completo and tipo_completo != 'Desconhecido' else ""
        
        if sucesso:
            print(f"[{processados}/{total_para_enviar}] ✅ OK{fase_info} - {chave_formatada}{tipo_info}")
            return ('sucesso', xml_info, msg)
        else:
            print(f"[{processados}/{total_para_enviar}] ❌ ERRO{fase_info} - {chave_formatada}{tipo_info}")
            print(f"         {msg}")
            return ('erro_recuperavel' if recuperavel else 'erro', xml_info, msg)
    
    # Fase 1: Sequencial (warm-up inicial)
    fase1_qtd = min(WARM_UP_FASE1_QTD, len(xmls_para_enviar))
    if fase1_qtd > 0:
        print(f"── Fase 1: Enviando {fase1_qtd} XMLs sequencialmente ──")
        for xml_info in xmls_para_enviar[:fase1_qtd]:
            resultado, xml, msg = enviar_com_progresso(xml_info, " [F1]")
            if resultado == 'sucesso':
                enviados_sucesso.append(xml)
            elif resultado == 'erro_recuperavel':
                erros_recuperaveis.append(xml)
            else:
                erros_envio.append({'xml': xml, 'erro': msg})
            time.sleep(WARM_UP_DELAY)
    
    restantes = xmls_para_enviar[fase1_qtd:]
    
    # Fase 2: Poucas threads
    fase2_qtd = min(WARM_UP_FASE2_QTD, len(restantes))
    if fase2_qtd > 0:
        print(f"\n── Fase 2: Enviando {fase2_qtd} XMLs com {WARM_UP_FASE2_THREADS} threads ──")
        with ThreadPoolExecutor(max_workers=WARM_UP_FASE2_THREADS) as executor:
            futures = {executor.submit(enviar_com_progresso, xml, " [F2]"): xml for xml in restantes[:fase2_qtd]}
            for future in as_completed(futures):
                resultado, xml, msg = future.result()
                if resultado == 'sucesso':
                    enviados_sucesso.append(xml)
                elif resultado == 'erro_recuperavel':
                    erros_recuperaveis.append(xml)
                else:
                    erros_envio.append({'xml': xml, 'erro': msg})
    
    restantes = restantes[fase2_qtd:]
    
    # Fase 3: Mais threads
    fase3_qtd = min(WARM_UP_FASE3_QTD, len(restantes))
    if fase3_qtd > 0:
        print(f"\n── Fase 3: Enviando {fase3_qtd} XMLs com {WARM_UP_FASE3_THREADS} threads ──")
        with ThreadPoolExecutor(max_workers=WARM_UP_FASE3_THREADS) as executor:
            futures = {executor.submit(enviar_com_progresso, xml, " [F3]"): xml for xml in restantes[:fase3_qtd]}
            for future in as_completed(futures):
                resultado, xml, msg = future.result()
                if resultado == 'sucesso':
                    enviados_sucesso.append(xml)
                elif resultado == 'erro_recuperavel':
                    erros_recuperaveis.append(xml)
                else:
                    erros_envio.append({'xml': xml, 'erro': msg})
    
    restantes = restantes[fase3_qtd:]
    
    # Fase 4: Todas as threads
    if restantes:
        print(f"\n── Fase 4: Enviando {len(restantes)} XMLs com {num_threads} threads ──")
        with ThreadPoolExecutor(max_workers=num_threads) as executor:
            futures = {executor.submit(enviar_com_progresso, xml, ""): xml for xml in restantes}
            for future in as_completed(futures):
                resultado, xml, msg = future.result()
                if resultado == 'sucesso':
                    enviados_sucesso.append(xml)
                elif resultado == 'erro_recuperavel':
                    erros_recuperaveis.append(xml)
                else:
                    erros_envio.append({'xml': xml, 'erro': msg})
    
    # Retentar erros recuperáveis
    if erros_recuperaveis:
        print(f"\n── Retentando {len(erros_recuperaveis)} XMLs com erro recuperável ──")
        time.sleep(2)  # Pausa antes de retentar
        
        for xml_info in erros_recuperaveis:
            resultado, xml, msg = enviar_com_progresso(xml_info, " [RETRY]")
            if resultado == 'sucesso':
                enviados_sucesso.append(xml)
            else:
                erros_envio.append({'xml': xml, 'erro': msg})
    
    tempo_total = time.time() - inicio_tempo
    
    # Excluir XMLs enviados com sucesso
    if excluir_enviados and enviados_sucesso:
        print(f"\n🗑️  Excluindo {len(enviados_sucesso)} XMLs enviados com sucesso...")
        excluidos = 0
        for xml_info in enviados_sucesso:
            try:
                os.remove(xml_info['caminho'])
                excluidos += 1
            except:
                pass
        print(f"   ✅ {excluidos} arquivo(s) excluído(s)")
    
    # Resumo final
    print(f"\n{'='*60}")
    print("📊 RESUMO FINAL DO UPLOAD")
    print("="*60)
    print(f"   📁 Total de XMLs encontrados: {len(xmls_validos)}")
    print(f"   Ja existentes no SIEG: {len(xmls_ja_existentes)} (verificacao previa desativada)")
    print(f"   ✅ Enviados com sucesso: {len(enviados_sucesso)}")
    print(f"   ❌ Erros: {len(erros_envio)}")
    print(f"   ⏱️  Tempo total: {tempo_total:.2f} segundos")
    if total_para_enviar > 0 and tempo_total > 0:
        print(f"   🚀 Velocidade: {len(enviados_sucesso)/tempo_total:.2f} XMLs/segundo")
    
    if erros_envio:
        print(f"\n   ❌ Detalhes dos erros ({len(erros_envio)}):")
        for erro in erros_envio[:10]:
            print(f"      - {erro['xml']['nome'][:40]}...: {erro['erro'][:50]}")
        if len(erros_envio) > 10:
            print(f"      ... e mais {len(erros_envio) - 10} erro(s)")
    
    print("="*60)
    print()
    
    return {
        "total": len(xmls_validos),
        "enviados": len(enviados_sucesso),
        "existentes": len(xmls_ja_existentes),
        "erros": len(erros_envio)
    }

File: baixar-e-solicitar-xml-sefaz/TS/src-ts/upload/uploader.ts
import pLimit from "p-limit";
import { existsSync, readdirSync, readFileSync, rmSync, statSync } from "node:fs";
import { basename, join } from "node:path";
import { PATHS, SIEG_API_KEY } from "../core/config.js";
import type { UploadResult } from "../types.js";
import { sleep } from "../utils/retry.js";
import { enviarXml, verificarEventoExisteComRetry, verificarXmlExisteComRetry } from "./sieg-api.js";
import {
  classificarXmlSieg,
  extrairChaveAcesso,
  extrairTipoEvento,
  identificarTipoXml,
  obterTipoCompletoNota,
  obterXmlTypeSieg,
  type XmlCategoriaSieg,
  validarXmlDetalhado,
} from "./utils.js";

interface XmlInfo {
  caminho: string;
  nome: string;
  conteudo: string;
  chave?: string;
  tipo?: string;
  categoria: XmlCategoriaSieg;
  tipoEvento?: number;
  xmlTypeSieg: number;
}

interface UploadState {
  apiKey: string;
  total: number;
  iniciados: number;
  concluidos: number;
  retryIniciados: number;
  retryTotal: number;
  excluirEnviados: boolean;
  excluidos: number;
  enviadosSucesso: XmlInfo[];
  errosEnvio: Array<{ xml: XmlInfo; erro: string }>;
  errosRecuperaveis: XmlInfo[];
  errosExclusao: Array<{ xml: XmlInfo; erro: string }>;
  confirmadosSieg: XmlInfo[];
  eventosConfirmadosSieg: XmlInfo[];
  semValidacaoFinalSieg: XmlInfo[];
  naoConfirmadosSieg: XmlInfo[];
}

const WARM_UP_FASE1_QTD = 5;
const WARM_UP_FASE1_THREADS = 1;
const WARM_UP_FASE2_QTD = 15;
const WARM_UP_FASE2_THREADS = 5;
const WARM_UP_FASE3_QTD = 30;
const WARM_UP_FASE3_THREADS = 10;
const WARM_UP_DELAY_MS = 200;
const VERIFICACAO_FINAL_THREADS = 5;
export const NUM_THREADS_PADRAO = 20;

function listarXmlsRecursivo(pasta: string): string[] {
  const result: string[] = [];
  if (!existsSync(pasta)) {
    return result;
  }
  for (const item of readdirSync(pasta)) {
    const path = join(pasta, item);
    const stats = statSync(path);
    if (stats.isDirectory()) {
      result.push(...listarXmlsRecursivo(path));
    } else if (stats.isFile() && item.toLowerCase().endsWith(".xml")) {
      result.push(path);
    }
  }
  return result;
}

function carregarXmlsValidos(paths: string[]): { validos: XmlInfo[]; invalidos: Array<{ caminho: string; erro: string }> } {
  const xmlsValidos: XmlInfo[] = [];
  const xmlsInvalidos: Array<{ caminho: string; erro: string }> = [];
  paths.forEach((caminho, index) => {
    console.log(`[VALIDANDO ${index + 1}/${paths.length}] ${basename(caminho)}`);
    try {
      const conteudo = readFileSync(caminho, "utf8");
      const validacao = validarXmlDetalhado(conteudo);
      if (!validacao.valido) {
        xmlsInvalidos.push({ caminho, erro: validacao.erro });
        return;
      }
      const chave = extrairChaveAcesso(conteudo);
      xmlsValidos.push({
        caminho,
        nome: basename(caminho),
        conteudo,
        chave,
        tipo: identificarTipoXml(conteudo),
        categoria: classificarXmlSieg(conteudo),
        tipoEvento: extrairTipoEvento(conteudo),
        xmlTypeSieg: obterXmlTypeSieg(conteudo),
      });
    } catch (error) {
      xmlsInvalidos.push({ caminho, erro: error instanceof Error ? error.message : String(error) });
      // Mantem o comportamento tolerante para arquivos ilegiveis.
    }
  });
  return { validos: xmlsValidos, invalidos: xmlsInvalidos };
}

function formatarIdentificacaoXml(xmlInfo: XmlInfo): string {
  const prefixo = xmlInfo.tipo === "CTe" ? "C" : "N";
  const chave = xmlInfo.chave && xmlInfo.chave.length === 44 ? `${prefixo}_${xmlInfo.chave}` : `${xmlInfo.nome.slice(0, 45)}...`;
  const tipoCompleto = obterTipoCompletoNota(xmlInfo.conteudo);
  const tipoInfo = tipoCompleto && tipoCompleto !== "Desconhecido" ? ` (${tipoCompleto})` : "";
  return `${chave}${tipoInfo}`;
}

function excluirXmlEnviado(xmlInfo: XmlInfo, state: UploadState): void {
  if (!state.excluirEnviados) {
    return;
  }

  try {
    rmSync(xmlInfo.caminho, { force: true });
    state.excluidos += 1;
    console.log(`         [EXCLUIDO] ${xmlInfo.nome}`);
  } catch (error) {
    const mensagem = error instanceof Error ? error.message : String(error);
    state.errosExclusao.push({ xml: xmlInfo, erro: mensagem });
    console.log(`         [AVISO] Falha ao excluir ${xmlInfo.nome}: ${mensagem}`);
  }
}

async function enviarLote(
  xmls: XmlInfo[],
  threads: number,
  faseInfo: string,
  state: UploadState,
  options: { retry?: boolean } = {},
): Promise<void> {
  const limit = pLimit(Math.max(1, threads));
  const isRetry = options.retry ?? false;

  await Promise.all(
    xmls.map((xmlInfo) =>
      limit(async () => {
        const identificacao = formatarIdentificacaoXml(xmlInfo);
        if (isRetry) {
          state.retryIniciados += 1;
          console.log(`[ENVIANDO RETRY ${state.retryIniciados}/${state.retryTotal}]${faseInfo} - ${identificacao}`);
        } else {
          state.iniciados += 1;
          console.log(`[ENVIANDO ${state.iniciados}/${state.total}]${faseInfo} - ${identificacao}`);
        }

        const [sucesso, msg, recuperavel] = await enviarXml(xmlInfo.conteudo, state.apiKey, undefined, true);

        if (sucesso) {
          state.concluidos += 1;
          console.log(`[OK ${state.concluidos}/${state.total}]${faseInfo} - ${identificacao}`);
          state.enviadosSucesso.push(xmlInfo);
          return;
        }

        if (recuperavel && !isRetry) {
          console.log(`[RETRY ${state.concluidos}/${state.total}]${faseInfo} - ${identificacao}`);
          console.log(`         ${msg}`);
          state.errosRecuperaveis.push(xmlInfo);
          return;
        }

        state.concluidos += 1;
        console.log(`[ERRO ${state.concluidos}/${state.total}]${faseInfo} - ${identificacao}`);
        console.log(`         ${msg}`);
        state.errosEnvio.push({ xml: xmlInfo, erro: msg });
      }),
    ),
  );
}

async function validarEnviadosNoSieg(state: UploadState): Promise<void> {
  if (!state.enviadosSucesso.length) {
    return;
  }

  console.log("\n" + "=".repeat(60));
  console.log("Validando XMLs enviados no SIEG...");
  console.log("=".repeat(60));

  const limit = pLimit(VERIFICACAO_FINAL_THREADS);
  let iniciados = 0;

  await Promise.all(
    state.enviadosSucesso.map((xmlInfo) =>
      limit(async () => {
        const indice = ++iniciados;
        const identificacao = formatarIdentificacaoXml(xmlInfo);
        console.log(`[VALIDANDO SIEG ${indice}/${state.enviadosSucesso.length}] ${identificacao}`);

        if (xmlInfo.categoria === "inutilizacao") {
          registrarSemValidacaoFinal(xmlInfo, state, "Inutilizacao aceita no upload; sem endpoint publico de validacao final");
          return;
        }

        if (!xmlInfo.chave || xmlInfo.chave.length !== 44) {
          registrarSemValidacaoFinal(xmlInfo, state, "XML aceito no upload, mas sem chave de acesso para validacao final");
          return;
        }

        const xmlTypeSieg = xmlInfo.xmlTypeSieg;
        if (obterTipoCompletoNota(xmlInfo.conteudo) === "Desconhecido") {
          console.log("         [AVISO] Tipo de XML nao identificado; validando no SIEG como NF-e (xmlType=1)");
        }

        if (xmlInfo.categoria === "evento") {
          if (!xmlInfo.tipoEvento || !Number.isFinite(xmlInfo.tipoEvento)) {
            registrarSemValidacaoFinal(xmlInfo, state, "Evento aceito no upload, mas sem tpEvento para validacao final");
            return;
          }

          const eventoConfirmado = await verificarEventoExisteComRetry(
            xmlInfo.chave,
            xmlInfo.tipoEvento,
            state.apiKey,
            undefined,
            undefined,
            xmlTypeSieg,
          );
          if (eventoConfirmado) {
            state.eventosConfirmadosSieg.push(xmlInfo);
            console.log("         [SIEG OK] Evento confirmado no SIEG");
            excluirXmlEnviado(xmlInfo, state);
          } else {
            state.naoConfirmadosSieg.push(xmlInfo);
            console.log("         [AVISO] Evento enviado, mas ainda nao foi confirmado no SIEG na validacao final");
          }
          return;
        }

        const confirmado = await verificarXmlExisteComRetry(xmlInfo.chave, state.apiKey, undefined, undefined, xmlTypeSieg);
        if (confirmado) {
          state.confirmadosSieg.push(xmlInfo);
          console.log("         [SIEG OK] XML confirmado no SIEG");
          excluirXmlEnviado(xmlInfo, state);
        } else {
          state.naoConfirmadosSieg.push(xmlInfo);
          console.log("         [AVISO] XML enviado, mas ainda nao foi confirmado no SIEG na validacao final");
        }
      }),
    ),
  );
}

function registrarSemValidacaoFinal(xmlInfo: XmlInfo, state: UploadState, mensagem: string): void {
  state.semValidacaoFinalSieg.push(xmlInfo);
  console.log(`         [SIEG SEM VALIDACAO FINAL] ${mensagem}`);
  excluirXmlEnviado(xmlInfo, state);
}

export async function enviarAutomatico(options: {
  pasta?: string;
  excluirEnviados?: boolean;
  numThreads?: number;
} = {}): Promise<UploadResult> {
  const pasta = options.pasta ?? PATHS.downloadsDir;
  const excluirEnviados = options.excluirEnviados ?? true;
  const numThreads = options.numThreads ?? NUM_THREADS_PADRAO;

  console.log("=".repeat(60));
  console.log("MODO AUTOMATICO - Envio de XMLs para SIEG (TS/Bun)");
  console.log("=".repeat(60));
  console.log(`\nPasta: ${pasta}`);
  console.log(`Threads max: ${numThreads}`);
  console.log(`Excluir apos envio: ${excluirEnviados ? "Sim" : "Nao"}\n`);

  if (!existsSync(pasta)) {
    console.log(`\nERRO: Pasta nao encontrada: ${pasta}`);
    return { erro: "Pasta nao encontrada", total: 0, enviados: 0, existentes: 0, erros: 0 };
  }

  const apiKey = SIEG_API_KEY;
  if (!apiKey) {
    console.log("\nERRO: API Key nao configurada no .env");
    return { erro: "API Key nao configurada", total: 0, enviados: 0, existentes: 0, erros: 0 };
  }

  console.log("\n" + "=".repeat(60));
  console.log("Buscando arquivos XML...");
  console.log("=".repeat(60));
  const xmlsEncontrados = listarXmlsRecursivo(pasta);
  if (!xmlsEncontrados.length) {
    console.log("[INFO] Nenhum arquivo XML encontrado na pasta.");
    return { total: 0, enviados: 0, existentes: 0, erros: 0 };
  }
  console.log(`Encontrados: ${xmlsEncontrados.length} arquivo(s) XML`);

  console.log("\n" + "=".repeat(60));
  console.log("Lendo e validando XMLs...");
  console.log("=".repeat(60));
  const { validos: xmlsValidos, invalidos: xmlsInvalidos } = carregarXmlsValidos(xmlsEncontrados);
  console.log(`XMLs validos: ${xmlsValidos.length}`);
  if (xmlsInvalidos.length) {
    console.log(`XMLs invalidos/ilegiveis: ${xmlsInvalidos.length}`);
    for (const invalido of xmlsInvalidos.slice(0, 5)) {
      console.log(`   [INVALIDO] ${basename(invalido.caminho)} - ${invalido.erro}`);
    }
    if (xmlsInvalidos.length > 5) {
      console.log(`   ... ${xmlsInvalidos.length - 5} outro(s) XML(s) invalido(s) omitido(s)`);
    }
  }
  if (!xmlsValidos.length) {
    return { total: 0, enviados: 0, existentes: 0, erros: 0 };
  }

  console.log("\n" + "=".repeat(60));
  console.log("Preparando XMLs para envio direto ao SIEG...");
  console.log("=".repeat(60));
  console.log("Consulta previa de existencia desativada; todos os XMLs validos serao enviados.");

  const state: UploadState = {
    apiKey,
    total: xmlsValidos.length,
    iniciados: 0,
    concluidos: 0,
    retryIniciados: 0,
    retryTotal: 0,
    excluirEnviados,
    excluidos: 0,
    enviadosSucesso: [],
    errosEnvio: [],
    errosRecuperaveis: [],
    errosExclusao: [],
    confirmadosSieg: [],
    eventosConfirmadosSieg: [],
    semValidacaoFinalSieg: [],
    naoConfirmadosSieg: [],
  };
  const inicioTempo = Date.now();

  const fase1 = xmlsValidos.slice(0, WARM_UP_FASE1_QTD);
  if (fase1.length) {
    console.log(`\n-- Fase 1: Enviando ${fase1.length} XMLs sequencialmente --`);
    for (const xml of fase1) {
      await enviarLote([xml], WARM_UP_FASE1_THREADS, " [F1]", state);
      await sleep(WARM_UP_DELAY_MS);
    }
  }

  let restantes = xmlsValidos.slice(fase1.length);
  const fase2 = restantes.slice(0, WARM_UP_FASE2_QTD);
  if (fase2.length) {
    console.log(`\n-- Fase 2: Enviando ${fase2.length} XMLs com ${WARM_UP_FASE2_THREADS} threads --`);
    await enviarLote(fase2, WARM_UP_FASE2_THREADS, " [F2]", state);
  }

  restantes = restantes.slice(fase2.length);
  const fase3 = restantes.slice(0, WARM_UP_FASE3_QTD);
  if (fase3.length) {
    console.log(`\n-- Fase 3: Enviando ${fase3.length} XMLs com ${WARM_UP_FASE3_THREADS} threads --`);
    await enviarLote(fase3, WARM_UP_FASE3_THREADS, " [F3]", state);
  }

  restantes = restantes.slice(fase3.length);
  if (restantes.length) {
    console.log(`\n-- Fase 4: Enviando ${restantes.length} XMLs com ${numThreads} threads --`);
    await enviarLote(restantes, numThreads, "", state);
  }

  if (state.errosRecuperaveis.length) {
    const retryList = [...state.errosRecuperaveis];
    state.errosRecuperaveis.length = 0;
    state.retryIniciados = 0;
    state.retryTotal = retryList.length;
    console.log(`\n-- Retentando ${retryList.length} XMLs com erro recuperavel --`);
    await sleep(2000);
    await enviarLote(retryList, 1, " [RETRY]", state, { retry: true });
  }

  await validarEnviadosNoSieg(state);

  const tempoTotal = (Date.now() - inicioTempo) / 1000;
  console.log("\n" + "=".repeat(60));
  console.log("RESUMO FINAL DO UPLOAD");
  console.log("=".repeat(60));
  console.log(`   Total de XMLs encontrados: ${xmlsValidos.length}`);
  console.log("   Ja existentes no SIEG: 0 (verificacao previa desativada)");
  console.log(`   Enviados com sucesso: ${state.enviadosSucesso.length}`);
  console.log(`   Confirmados no SIEG: ${state.confirmadosSieg.length}`);
  console.log(`   Eventos confirmados no SIEG: ${state.eventosConfirmadosSieg.length}`);
  console.log(`   Sem validacao final no SIEG: ${state.semValidacaoFinalSieg.length}`);
  console.log(`   Nao confirmados no SIEG: ${state.naoConfirmadosSieg.length}`);
  console.log(`   Erros: ${state.errosEnvio.length}`);
  console.log(`   XMLs excluidos: ${state.excluidos}`);
  console.log(`   Falhas ao excluir: ${state.errosExclusao.length}`);
  console.log(`   Tempo total: ${tempoTotal.toFixed(2)} segundos`);
  console.log("=".repeat(60));

  return {
    total: xmlsValidos.length,
    enviados: state.enviadosSucesso.length,
    existentes: 0,
    erros: state.errosEnvio.length,
    confirmadosSieg: state.confirmadosSieg.length,
    eventosConfirmadosSieg: state.eventosConfirmadosSieg.length,
    semValidacaoFinalSieg: state.semValidacaoFinalSieg.length,
    naoConfirmadosSieg: state.naoConfirmadosSieg.length,
  };
}
